core/server/proxy/src/main/java/alluxio/proxy/s3/signature/AwsSignatureProcessor.java

```diff
@@ -0,0 +1,229 @@
+ **/
+
+public class AwsSignatureProcessor {
+  private static final Logger LOG =
+            LoggerFactory.getLogger(AwsSignatureProcessor.class);
+  private static final String AUTHORIZATION = "Authorization";
+
+  private ContainerRequestContext mContext;
+
+  /**
+   * Create a new {@link AwsSignatureProcessor}.
+   *
+   * @param context ContainerRequestContext
+   */
+  public AwsSignatureProcessor(ContainerRequestContext context) {
+    mContext = context;
+  }
+
+  /**
+   * Extract signature info from request.
+   * @return SignatureInfo
+   * @throws S3Exception
+   */
+  public SignatureInfo parseSignature() throws S3Exception {
+    LowerCaseKeyStringMap headers =
+        LowerCaseKeyStringMap.fromHeaderMap(mContext.getHeaders());
+
+    String authHeader = headers.get(AUTHORIZATION);
+    String dateHeader = headers.get(X_AMZ_DATE);
+    Map<String, String> queryParameters = StringToSignProducer.fromMultiValueToSingleValueMap(
+            mContext.getUriInfo().getQueryParameters());
+
+    SignatureInfo signatureInfo = null;
+    if ((signatureInfo =
+            AwsAuthV2HeaderParserUtils.parseSignature(authHeader)) != null
+        || (signatureInfo =
+            AwsAuthV4HeaderParserUtils.parseSignature(authHeader, dateHeader)) != null
+        || (signatureInfo =
+            AwsAuthV4QueryParserUtils.parseSignature(queryParameters)) != null) {
+      return signatureInfo;
+    } else {
+      return new SignatureInfo(
+              SignatureInfo.Version.NONE,
+              "", "", "", "", "", "", "", false
+      );
+    }
+  }
+
+  /**
+   * Convert SignatureInfo to AwsAuthInfo.
+   * @return AwsAuthInfo
+   * @throws S3Exception
+   */
+  public AwsAuthInfo getAuthInfo() throws S3Exception {
+    try {
+      SignatureInfo signatureInfo = parseSignature();
+      String stringToSign = "";
+      if (signatureInfo.getVersion() == SignatureInfo.Version.V4) {
+        stringToSign =
+                StringToSignProducer.createSignatureBase(signatureInfo, mContext);
+      }
+      String awsAccessId = signatureInfo.getAwsAccessId();
+      // ONLY validate aws access id when needed.
+      if (StringUtils.isEmpty(awsAccessId)) {
+        LOG.debug("Malformed s3 header. awsAccessID is empty");
+        throw new S3Exception("awsAccessID is empty", S3ErrorCode.ACCESS_DENIED_ERROR);
+      }
+
+      return new AwsAuthInfo(awsAccessId,
+              stringToSign,
+              signatureInfo.getSignature()
+              );
+    } catch (S3Exception ex) {
+      LOG.debug("Error during signature parsing: ", ex);
+      throw ex;
+    } catch (Exception e) {
+      // For any other critical errors during object creation throw Internal
+      // error.
+      LOG.debug("Error during signature parsing: ", e);
+      throw new S3Exception(e, "Context is invalid", S3ErrorCode.INTERNAL_ERROR);
+    }
+  }
+
+  /**
+   * A simple map which forces lower case key usage.
```

Comment:
If your goal is to just force lower case string key, It might be easier to just define a lowercase string class and override the comparator. That would be less code to write than a map implementation.

Reply:
finish.